src/ethereum/event-fetcher-lookahead.ts

```diff
@@ -0,0 +1,92 @@
+import _ from 'lodash';
+import { EventFetcher } from './event-fetcher';
+import { EventData } from 'web3-eth-contract';
+import { EventName } from './types';
+import { EthereumReader } from './ethereum-reader';
+import * as Logger from '../logger';
+
+export interface AutoscaleOptions {
+  initialPageSize: number; // the initial window size (max blocks in batch)
+  maxPageSize: number; // max possible window size (since it's autoscaling)
+  minPageSize: number; // min possible window size (since it's autoscaling)
+  pageGrowFactor: number; // by how much should the window grow when attempting to grow
+  pageGrowAfter: number; // after how many consecutive successes should we attempt to grow
+  pageShrinkFactor: number; // by how much should the window shrink after a failure
+}
+
+// more efficient fetcher that supports lookahead
+// works with constant page size, which needs to be improved
+export class LookaheadEventFetcher extends EventFetcher {
+  protected lookAhead: EventData[] = [];
+  protected lookAheadFromBlock = 0;
+  protected lookAheadToBlock = 0;
+  protected currentPageSize = 0;
+  protected autoscale: AutoscaleOptions;
+  protected autoscaleStreak = 0;
+  public autoscaleConsecutiveFailures = 0;
```

Comment:
> I like the fetcher, because it gives a nice foundation for extending it to a multi-page fetcher (get the next page while the first is being read). But I don't really see the improvement over the bulk fetcher - you could have just replaced the `getPastEvents` call in the old bulk fetcher with your new `downloadNewPage` - i think the two fetchers are pretty equivalent in all other respects

I originally started from the original LookAheadFetcher, as you recall it was much more complicated since it was downloading more lookahead with the background promise. I tried to add the autoscale feature to it, but the code was too hard to follow so I simplified and simplified and ended up with this :)

Reply:
> never read. is this field exposed somewhere? is it exposed in the status page?

I stole the autoscale code from the reward distributor (it doesn't have a fetcher but it did have autoscale). In there we expose the number of consecutive failures as a public field to the caller so they could tell the difference from autoscale failures (too big thus failing) which are ok, to actual network failures (like infura is down). If the consecutive failures gets too big, you're expected to stop trying and rest a while. This behavior is not needed for our service. I kept the code because I plan to steal the entire fetcher back to rewards distributor so it can track registry changes as well and have a single codebase instead of two completely different implementations like today.